### ISSUE TITLE
On macOS Catalina unsigned binaries blocked by Gatekeeper e.g. kernel-service.dart.snapshot, analysis_server.dart.snapshot

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#36714]
For macOS Catalina, Apple has changed the way Gatekeeper verifies binaries downloaded from the internet ([reference](https://eclecticlight.co/2019/06/12/grokking-gatekeeper-in-catalina/)), and apparently disabled the "open anywhere" option from System Preferences. Currently, using the flutter tool is not possible unless the user runs `sudo spctl --master-disable` from the command line [reference](http://osxdaily.com/2015/05/04/disable-gatekeeper-command-line-mac-osx/).

The following commands are flagged:

1. `flutter/bin/cache/dart-sdk/bin/dart` binary
1. `flutter/bin/cache/dart-sdk/bin/snapshots/*.dart.snapshot`

Here is some log output when issuing command `flutter channel master`:

```
flutter@macN1 flutter % bin/flutter channel master
Switching to flutter channel 'master'...
../../third_party/dart/runtime/bin/snapshot_utils.cc: 149: error: Failed to memory map snapshot: /Users/flutter/flutter/bin/cache/dart-sdk/bin/snapshots/kernel-service.dart.snapshot
 
version=2.4.0 (Wed Jun 19 11:53:45 2019 +0200) on "macos_x64"
thread=5891, isolate=(null)(0x0)
  pc 0x000000010bd945d5 fp 0x0000700006ee7c40 dart::Profiler::DumpStackTrace(void*)
  pc 0x000000010badfdc2 fp 0x0000700006ee7d20 dart::Assert::Fail(char const*, ...)
  pc 0x000000010bac2dfb fp 0x0000700006ee7dd0 dart::bin::Snapshot::TryReadAppSnapshot(char const*)
  pc 0x000000010bac7eea fp 0x0000700006ee7e50 dart::bin::CreateIsolateAndSetup(char const*, char const*, char const*, char const*, Dart_IsolateFlags*, void*, char**)
  pc 0x000000010bcdba48 fp 0x0000700006ee7f00 dart::RunKernelTask::Run()
  pc 0x000000010be36a0d fp 0x0000700006ee7f30 dart::ThreadPool::Worker::Loop()
  pc 0x000000010be36880 fp 0x0000700006ee7f70 dart::ThreadPool::Worker::Main(unsigned long)
  pc 0x000000010bd90fc1 fp 0x0000700006ee7fb0 dart::ThreadStart(void*)
  pc 0x00007fff73470cce fp 0x0000700006ee7fd0 _pthread_start
  pc 0x00007fff7346d72b fp 0x0000700006ee7ff0 thread_start
-- End of DumpStackTrace
bin/flutter: line 183:  8210 Abort trap: 6           "$DART" --packages="$FLUTTER_TOOLS_DIR/.packages" $FLUTTER_TOOL_ARGS "$SNAPSHOT_PATH" "$@"
```

Doctor output:

```
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel master, v1.8.2-pre.133, on Mac OS X 10.15 19A512f, locale en-US)
    ✗ Downloaded executables cannot execute on host.
      See https://github.com/flutter/flutter/issues/6207 for more information

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
[✓] Chrome - develop for the web
[!] Android Studio (version 3.4)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[!] VS Code (version 1.36.1)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[✓] Connected device (3 available)

! Doctor found issues in 3 categories.
flutter@macN1 flutter % flutter doctor -v
[!] Flutter (Channel master, v1.8.2-pre.133, on Mac OS X 10.15 19A512f, locale en-US)
    • Flutter version 1.8.2-pre.133 at /Users/flutter/flutter
    • Framework revision d0238c711e (6 hours ago), 2019-07-22 12:14:27 -0400
    • Engine revision 849fe73170
    • Dart version 2.5.0 (build 2.5.0-dev.1.0 a089199b93)
    ✗ Downloaded executables cannot execute on host.
      See https://github.com/flutter/flutter/issues/6207 for more information

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.1)
    • Android SDK at /Users/flutter/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
    • Xcode at /Users/flutter/Downloads/Xcode-beta.app/Contents/Developer
    • Xcode 11.0, Build version 11M374r
    • CocoaPods version 1.7.5

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[!] VS Code (version 1.36.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS      • macOS                                    • darwin-x64     • Mac OS X 10.15 19A512f
    • Chrome     • chrome                                   • web-javascript • Google Chrome 75.0.3770.142

! Doctor found issues in 3 categories.
```

**Comment by zanderso** [comment]
It looks like running from an app-jit snapshot may generally have issues on Catalina, not just in the context of the Flutter tool. @a-siva @rmacnak-google

**Comment by rmacnak-google** [comment]
These snapshots are produced during Flutter engine builds, so any signing would need to happen there.

I don't know if Gatekeeper allows for signatures that aren't on Mach-O files. If it doesn't, the only thing would can do here to downgrade from AppJIT snapshots to plain App snapshot with the GN flag `dart_snapshot_kind = "app"`.

@chinmaygarde

**Comment by christopherfujino** [comment]
the dart binary itself is also flagged by Gatekeeper.

**Comment by rmacnak-google** [comment]
Then how did you get Dart binary to run up to the mmap backtrace? Did you get a dialog for the Dart binary with the choice to run it anyway, with Gatekeeper not giving that choice for the snapshot?

**Comment by christopherfujino** [comment]
@rmacnak-google more or less. You can see my step-by-step journal [here](https://github.com/flutter/flutter/wiki/State-of-Catalina-Support).

```
3. When I enter the command flutter/bin/flutter doctor, OS X says “dart” can’t be opened because it was not downloaded from the App Store
  i. After going to Settings -> Security & Privacy -> Open Anyway, I get another dialog: dart is a Unix app downloaded from the Internet. Are you sure you want to open it?
  ii. On clicking “open”, a new “Terminal” window was opened, and then it executes just dart, without any arguments, and then the terminal session exits.
4. When I re-issue the same command, `flutter/bin/flutter doctor`, this time the actual flutter tool is run, however a new dialog is opened: `kernel-service.dart.snapshot` can’t be opened because Apple cannot check it for malicious software.
  i. When I go to Settings -> Security & Privacy -> Open Anyway, nothing happens. The text next to the button says: kernel-service.dart.snapshot was blocked from opening because it is not from an identified developer.
  ii. Re-running the command results in no change.
```

It was at this point I globally disabled gatekeeper (`sudo spctl --master-disable`) because I could not proceed otherwise.

**Comment by zanderso** [comment]
@a-siva

**Comment by a-siva** [comment]
Here is some documentation of a new option to mmap called MAP_JIT which was introduced recently https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-jit

**Comment by zanderso** [comment]
@christopherfujino do we also need to sign flutter_tester, and the iOS tools we build?

**Comment by a-siva** [comment]
@chinmaygarde did a quick test on a small test program with the MAP_JIT flag set and entitlement  Allow Unsigned Execution Memory turned on during signing and this worked.

Next steps are
- Dart VM should add the flag MAP_JIT to the mmap call when kReadExecute mode is used
- The Dart SDK build on macos and the Flutter engine builds should sign/notarize all the executables, In Addition the entitlements 'Allow Unsigned Execution Memory' and 'Allow Execution of JIT-compiled code' should be set for the Dart binary.

**Comment by a-siva** [comment]
@athomas

**Comment by christopherfujino** [comment]
@zanderso yes and yes. Both can use the same CRON process.

**Comment by a-siva** [comment]
Here is a relevant issue from chromium regarding this https://bugs.chromium.org/p/chromium/issues/detail?id=850193 and
a CL https://chromium-review.googlesource.com/c/chromium/src/+/1613809 that landed for chromium.

**Comment by christopherfujino** [comment]
This was fixed in https://github.com/dart-lang/sdk/commit/ec2d06d4b9f4f0accad2b4aa841499e8e93cd70b and by signing & notarizing binaries.

**Comment by rotter** [comment]
I have just downloaded flutter and got the mentioned message: "dart" cant be opened because Apple cannot check it for malicious software. Im not sure if that message shouldnt be shown due to this issue being closed, or if it's unrelated, but thought I should mention.

**Comment by rotter** [comment]
I tried the master-disable command, re-ran "flutter doctor" and it advanced a bit more and I got a second error: “idevice_id” cannot be opened because the developer cannot be verified.

I heard about flutter on a Google event a few months ago and was gonna check it now to see how stable it is, if it's worth using it for real software. Well, I guess I have the answer already

**Comment by christopherfujino** [comment]
@rotter what version did you download?

**Comment by rotter** [comment]
The version on https://flutter.dev/docs/get-started/install/macos :

flutter_macos_v1.9.1+hotfix.6-stable.zip

**Comment by christopherfujino** [comment]
This is a known issue that I'm working on, you can track it here https://github.com/flutter/flutter/issues/42969

**Comment by christopherfujino** [comment]
@rotter a workaround for the time being is to delete the cache, and have the tool repopulate the artifacts. This should work: ` rm -rf /path/to/flutter/directory/bin/cache/; flutter doctor -v`

**Comment by MaTToX3** [comment]
Still same error.

**Comment by eltriplet** [comment]
Try doing the following as a temporary workaround:

1. After getting the error, close the error dialog and go to 'System Preferences', then 'Security and Privacy'.
2. On the lower part of the first tab 'General' click on the button 'Open Anyway'
3. Re-run whatever threw the initial error message. A different dialog window will pop up, and click on 'Open' button.
4. Repeat these steps for each of the apps/processes that are flagged by GateKeeper.

**Comment by keitoro** [comment]
the above approach did also worked for me, I just stumbled upon the solution, but this is really not good and should be fixed soon.

**Comment by Kirito0512** [comment]
> The version on https://flutter.dev/docs/get-started/install/macos :
> 
> flutter_macos_v1.9.1+hotfix.6-stable.zip

my problem is same as you . my version is 1.9.1 hotfix6-stable too.  how do you solve this problem

**Comment by Kirito0512** [comment]
I find the solution , you can watch on [youtube](https://www.youtube.com/watch?v=qe2G3gXRMMw)

**Comment by rotter** [comment]
> @rotter a workaround for the time being is to delete the cache, and have the tool repopulate the artifacts. This should work: ` rm -rf /path/to/flutter/directory/bin/cache/; flutter doctor -v`

My friend tried installing Flutter yesterday on a Mac and he ran into the same problems, took us about an hour to sort everything out. Just wanted to point out this solution didnt work for him as the "cache" folder didnt even exist.

In the end the problem for him was that we had used the suggested "sudo spctl --master-disable" command. The problem is that with this command the GUI solution (which is to click on the "Allow" button on the System Preferences -> Security & Privacy -> General) wouldnt work, the Allow button wouldnt show up, because we disabled that underlying system, so you arrive on a dead end.

To fix you have to reenable it (sudo spctl --master-enable), then re-run flutter doctor, then go to System Preferences and click on Allow. Also we did a reboot at this point so a reboot might be needed as well.

**Comment by rotter** [comment]
PS: I dont understand why this issue is Closed, as it's clearly still very bugged (this problem happened with 100% of my sample data, and the solution is not straightforward).

Maybe it could get merged into another issue or something else, but not closed in my opinion.

**Comment by sorinteican** [comment]
Problem still not fixed v1.12.13+hotfix.7-stable, i don t want to google my way out of it, google should fix it...

**Comment by IchordeDionysos** [comment]
Just to track it here. My issue was this:

Before it had a folder structure like this:

```
dev
  bin (All my binaries)
    flutter (Flutter repo)
      bin (Flutter's binaries folder)
        flutter (Flutter Unix executable)
```
My path looked something like: 

```
PATH="... :/Users/user/dev/bin: ..."
```

I got it to work when I explicitly put the Flutter binaries folder into the PATH:

```
PATH="... :/Users/user/dev/bin/flutter/bin: ..."
```
